internal/jimm/juju/model.go

```diff
@@ -261,23 +261,49 @@ func (j *JujuManager) ModelInfo(ctx context.Context, user *openfga.User, mt name
 	}
 	defer api.Close()
 
-	mi := &jujuparams.ModelInfo{
-		UUID: mt.Id(),
-	}
-	if err := api.ModelInfo(ctx, mi); err != nil {
-		// If the model is not found or code not authorized on the backing controller then
-		// we delete the model from JIMM.
-		// Juju returns CodeAuthorized when the model is not found for this facade.
-		if errors.ErrorCode(err) == errors.CodeNotFound || errors.ErrorCode(err) == errors.CodeUnauthorized {
-			errDelete := j.deleteModel(ctx, mt)
-			if errDelete != nil {
-				return nil, errors.E(op, errDelete)
-			}
-		}
+	modelInfo, err := j.modelInfo(ctx, &m, api)
+	if err != nil {
 		return nil, errors.E(op, err)
 	}
 
-	return j.mergeModelInfo(ctx, user, mi, m)
+	return j.mergeModelInfo(ctx, user, modelInfo, m)
+}
+
+// modelInfo retrieves the model information from the controller and reacts
+// to the error to update JIMM's state.
+func (j *JujuManager) modelInfo(ctx context.Context, model *dbmodel.Model, api API) (*jujuparams.ModelInfo, error) {
+	modelInfo := &jujuparams.ModelInfo{UUID: model.UUID.String}
+	errFromAPI := api.ModelInfo(ctx, modelInfo)
+	var err error
+	switch model.MigrationMode {
+	case dbmodel.MigrationModeNone:
+		err = j.maybeCleanupModel(ctx, errFromAPI, model)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Mask the error here. If we fail to delete the model it would be weird to see `failed to delete model` in a call to `ModelInfo`. Not sure what message is appropriate though.
```suggestion
		if err != nil {
			return nil, errors.E(err, "internal server error")
		}
```

Reply:
i would rather mask it a level higher, like so:
```go
modelInfo, err := j.modelInfo(ctx, &m, api)
if err != nil {
zapctx.Error(ctx, "failed to get model info", zap.Error(err), zap.String("model", m.UUID.String))
return nil, errors.E(op, "internal server error")
}
```

Reply:
sike, we can't do that bucause we need to preserve the error code

Reply:
But you can't mask it higher because if the controller returned a normal error you want to display that? We know for sure we want to mask the error from `maybeCleanupModel`.

Reply:
we can do what you've suggested but it would be one of the very first place we do error masking.
Usually we don't do it, right?

Reply:
Ideally we'd wrap the "internal" error and let it bubble up, relying on an upper level not letting the client see more than they're allowed.

I think it may make most sense to wrap it in `internal server error` and log where it happens.

Reply:
i'm ok with starting now, but we don't usually do it

Reply:
I think the reason we don't normally mask the error is that we don't call these cleanup routines from within a handler that does something else.

Reply:
masking done